Civi/Api4/Generic/AbstractSaveAction.php

```diff
@@ -102,6 +105,16 @@ protected function validateValues() {
     if ($unmatched) {
       throw new \API_Exception("Mandatory values missing from Api4 {$this->getEntityName()}::{$this->getActionName()}: " . implode(", ", $unmatched), "mandatory_missing", ["fields" => $unmatched]);
     }
+    $e = new ValidateValuesEvent($this->getEntityName(), 'save', $this->records);
+    \Civi::dispatcher()->dispatch('civi.api4.validate', $e);
+    if (!empty($e->errors)) {
+      throw new \API_Exception(ts('Found %1 error(s) in submitted %2 record(s) of type "%3": %4', [
+        1 => count($e->errors),
+        2 => count(array_unique(array_column($e->errors, 'record'))),
+        3 => $this->getEntityName(),
+        4 => implode(', ', array_column($e->errors, 'message')),
+      ]));
+    }
```

Comment:
I think this whole block ought to be inside a `foreach` loop as the event is supposed to fire once per record not for an array of records. Also I think the second param (action name) shouldn't be 'save' but should be 'create' or 'update' depending on the presence of `$idField` like https://github.com/civicrm/civicrm-core/commit/20af3ca00dd30ae200d4dd22585bdef8dbaf1bd6#diff-ccb444858b7cfcd34c34786bd76003cdccad4c53fbadef2b89df5ba7a94e82f2R107